### PR TITLE
Snap 1523

### PR DESCRIPTION
**Pull request by rishitesh**
## Changes proposed in this pull request

This is a forked branch from SNAP-1454. Once 1454 is merged difference should be lot less.

a) Earlier we used a temp memory manager to record memory usage while GemXD is booting up. That caused problem if the data stored is more than memory allocated.
   As temp memory manager did not have any limits it used to allow recovering more than what can be accomodated. 
   Now changed temp managers to proper SnappyUnifiedMemoryManager , which will have all the memory limits in place. It can stop recovery after the threshold has been breached.

b) Recovery thread will now handle LowMemoryException from UMM. If it gets a LME it stops recovery process.

c) Alligned max storage fraction with eviction-heap-percentage. Could see some stable run by this.

d) No-Op MemoryManager now will delegate to Spark UnifiedMemoryManager. This is required for connector mode memory consumption by SnappyData operators.

## Patch testing

TODO precheckin

## ReleaseNotes.txt changes

NA

### CODE REVIEW

File: cluster/src/main/scala/org/apache/spark/memory/SnappyMemoryUtils.scala [hunk 1]
```diff
@@ -26,16 +25,16 @@ object SnappyMemoryUtils {
     *
     * @return
     */
-  def isCriticalUp(accountedMemory: Long): Boolean =
-    Option(GemFireCacheImpl.getInstance).exists { cache =>
-      cache.getResourceManager.getHeapMonitor.failMemoryRequest(accountedMemory)
-    }
+  def isCriticalUp(): Boolean =
+    Option(GemFireStore.getBootingInstance).exists(g => g.thresholdListener.isCritical)
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyMemoryUtils.scala]: _.thresholdListener.isCritical

File: cluster/src/main/scala/org/apache/spark/memory/SnappyMemoryUtils.scala [hunk 2]
```diff
@@ -26,16 +25,16 @@ object SnappyMemoryUtils {
...
-   * Checks whether GemFire eviction threshold is breached
-   * @return
-   */
-  def isEvictionUp: Boolean = {
+    * Checks whether GemFire eviction threshold is breached
+    *
+    * @return
+    */
+  def isEvictionUp: Boolean =
     Option(GemFireStore.getBootingInstance).exists(g => g.thresholdListener.isEviction)
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyMemoryUtils.scala]: _.thresholdListener.isEviction

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 3]
```diff
@@ -71,54 +74,77 @@ class SnappyUnifiedMemoryManager private[memory](
...
   private val onHeapStorageRegionSize = onHeapStorageMemoryPool.poolSize
 
-  private val maxHeapStorageSize = (maxHeapMemory *
-      conf.getDouble("spark.memory.storageMaxFraction", 0.9)).toLong
-  // TODO: [sumedh] Not being used?
-  val maxExecutionSize: Long = (maxHeapMemory * 0.75).toLong
+  val evictionFraction = SnappyUnifiedMemoryManager.getStorageEvictionPercent(conf)
+
+  private val maxHeapStorageSize =
+    (maxHeapMemory * evictionFraction * 0.01).toLong
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: change the name to evictionPercent

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 4]
```diff
@@ -71,54 +74,77 @@ class SnappyUnifiedMemoryManager private[memory](
...
-        // transfer the memory map from tempMemoryManager on first use
-        MemoryManagerCallback.tempMemoryManager.memoryForObject.map { entry =>
-          val (objectName, mode) = entry._1
-          acquireStorageMemoryForObject(objectName,
-            MemoryManagerCallback.storageBlockId, entry._2, mode, null,
-            shouldEvict = false)
-        }
-        MemoryManagerCallback.tempMemoryManager.memoryForObject.clear()
-        _memoryForObjectMap
+      if (memoryMap eq null) synchronized {
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: double synchronized?

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 5]
```diff
@@ -456,14 +481,21 @@ class SnappyUnifiedMemoryManager private[memory](
...
       if (!enoughMemory) {
 
         // return immediately for OFF_HEAP with shouldEvict=false
-        if (offHeapNoEvict) return false
+        if (offHeapNoEvict || tempManager) return false
 
-        if (!offHeap && SnappyMemoryUtils.isCriticalUp(getStoragePoolMemoryUsed(
-          MemoryMode.ON_HEAP) + getExecutionPoolUsedMemory(MemoryMode.ON_HEAP))) {
+        // @TODO Commenting the below code as it causes strange behaviour. Memory
+        // allocation fails here , but somehow the query does not get cancelled.
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Is this TODO valid? See no commented out code.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 6]
```diff
@@ -535,28 +568,34 @@ class SnappyUnifiedMemoryManager private[memory](
   override def releaseStorageMemoryForObject(objectName: String,
                                              numBytes: Long,
                                              memoryMode: MemoryMode): Unit = synchronized {
-    logDebug(s"releasing [SNAP] memory for $objectName = $numBytes")
-    if (memoryForObject.addTo(objectName, -numBytes) >= 0L) {
+    logDebug(s"releasing $managerId memory for $objectName = $numBytes")
+    if (memoryForObject.addTo(objectName -> memoryMode, -numBytes) != 0L) {
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Isn't >= 0 somewhat safer just in case there is an accounting mistake and results goes into -ve!

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 7]
```diff
@@ -535,28 +568,34 @@ class SnappyUnifiedMemoryManager private[memory](
...
       memoryForObject.removeLong(objectName)
     }
   }
 
-  override def releaseStorageMemory(numBytes: Long, memoryMode: MemoryMode): Unit =
+  override def releaseStorageMemory(numBytes: Long, memoryMode: MemoryMode): Unit = {
     releaseStorageMemoryForObject(SPARK_CACHE, numBytes, memoryMode)
+    logDebug(s"releasing $managerId memory for $SPARK_CACHE $numBytes")
+    if (memoryForObject.containsKey(SPARK_CACHE)) {
+      memoryForObject.addTo(SPARK_CACHE -> memoryMode, -numBytes)
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Key in containsKey is incorrect -- should be a pair with memoryMode. Better to create the key once and use in both.

A more efficient version that will avoid two key lookups for most cases:

```scala
val key = SPARK_CACHE -> memoryMode
if (memoryForObject.addTo(key, -numBytes) != 0L) {
  super.releaseStorageMemory(numBytes, memoryMode)
} else {
  // key was not present so remove
  memoryForObject.removeLong(key)
}
```

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 8]
```diff
@@ -217,18 +243,19 @@ class SnappyUnifiedMemoryManager private[memory](
...
-          val prev = memoryForObject.addTo(fromOwner, -totalSize)
+          val prev = memoryForObject.addTo(fromOwner -> mode, -totalSize)
           if (prev >= totalSize) {
-            memoryForObject.addTo(toOwner, totalSize)
+            memoryForObject.addTo(toOwner -> mode, totalSize)
           } else {
             // something went wrong with size accounting
-            memoryForObject.addTo(fromOwner, totalSize)
+            memoryForObject.addTo(fromOwner -> mode, totalSize)
             throw new IllegalStateException(
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: create key pair once instead of every time

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 9]
```diff
@@ -535,28 +568,34 @@ class SnappyUnifiedMemoryManager private[memory](
...
+    val bytesToBeFreed = memoryForObject.getLong(name -> memoryMode)
     val numBytes = Math.max(0, bytesToBeFreed - ignoreNumBytes)
-    logDebug(s"Dropping memory for $name = $numBytes (registered=$bytesToBeFreed)")
+    logDebug(s"Dropping $managerId memory for $name = $numBytes (registered=$bytesToBeFreed)")
 
     if (numBytes > 0) {
       super.releaseStorageMemory(numBytes, memoryMode)
-      memoryForObject.removeLong(name)
+      memoryForObject.removeLong(name -> memoryMode)
     }
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: can create key pair once

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 10]
```diff
@@ -580,6 +619,8 @@ object SnappyUnifiedMemoryManager extends Logging {
       math.max(getMaxHeapMemory / 20, 500L * 1024L * 1024L))
   }
 
+  private val DEFAULT_EVICTION_PERCENT = 80f
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: I think we can increase these defaults: critical=95%, eviction=90%

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 11]
```diff
@@ -635,6 +676,23 @@ object SnappyUnifiedMemoryManager extends Logging {
...
+    val cache = Misc.getGemFireCacheNoThrow
+    if (cache ne null) {
+      val thresholds = cache.getResourceManager.getHeapMonitor.getThresholds
+      if (thresholds.getEvictionThreshold > 0.1f) {
+        cache.getResourceManager.getHeapMonitor.getThresholds.getEvictionThreshold
+      } else {
+        DEFAULT_EVICTION_PERCENT
+      }
+    } else {
+      conf.getDouble("spark.testing.maxStorageFraction", DEFAULT_EVICTION_PERCENT).toFloat
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Property is a fraction which is getting mixed with percent. Use one consistent name to avoid confusion in this class -- perhaps fraction throughout since that is what Spark uses.

File: cluster/src/test/scala/io/snappydata/QueryTest.scala [hunk 12]
```diff
@@ -99,21 +98,4 @@ class QueryTest extends SnappyFunSuite {
...
-    session.sql("CREATE TABLE yes_with(device_id VARCHAR(200), " +
-        "sdk_version VARCHAR(200)) USING COLUMN OPTIONS(PARTITION_BY 'device_id')")
-    session.insert("yes_with", Row("id1", "v1"), Row("id1", "v2"),
-      Row("id2", "v1"), Row("id2", "v1"), Row("id2", "v3"))
-    val r = session.sql("select sdk_version, count(distinct device_id) from (" +
-        "select sdk_version,device_id from YES_WITH group by sdk_version, " +
-        "device_id) a group by sdk_version")
-    ColumnCacheBenchmark.collect(r,
-      Seq(Row("v1", 2), Row("v2", 1), Row("v3", 1)))
-  }
```
Comment (sumwale) [inline cluster/src/test/scala/io/snappydata/QueryTest.scala]: Why this change? Perhaps in resolving merge conflict -- revert.

File: core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala [hunk 13]
```diff
@@ -18,136 +18,86 @@ package org.apache.spark.memory
...
-    }
-    memoryForObject(key) += numBytes
-    true
+                                             blockId: BlockId,
+                                             numBytes: Long,
+                                             memoryMode: MemoryMode,
+                                             buffer: UMMMemoryTracker,
+                                             shouldEvict: Boolean): Boolean = {
+    logDebug(s"Acquiring DefaultManager meemory for $objectName $numBytes")
+    SparkEnv.get.memoryManager.acquireStorageMemory(blockId, numBytes, memoryMode)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala]: Correct the formatting in this class. The previous indentation was correct. Perhaps you need to set the codeStyleSettings.xml again with "Project" type in IDEA.

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala [hunk 14]
```diff
@@ -490,17 +534,18 @@ object CachedDataFrame
     if (dataLen == 0) return Iterator.empty
 
     val codec = CompressionCodec.createCodec(SparkEnv.get.conf)
-    val input = new Input(data, offset, dataLen)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Do you need allocateExecutionMemory calls for input data bytes too?

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala [hunk 15]
```diff
@@ -399,15 +407,51 @@ object CachedDataFrame
...
+        finalBuffer.flip
+        memSize = finalBuffer.limit()
+        // Ask UMM before getting the array to heap.
+        // Taking unroll memory as this memory is cleaned up on task completion.
+        // On connector mode also this should account to the overall memory usage.
+        // We will ensure that sufficient memory is available by reserving
+        // two times as Kryo serialization will expand its buffer accordingly.
+
+        if (!MemoryManagerCallback.memoryManager.
+          acquireStorageMemoryForObject(
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Why not allocate in execution memory?

File: core/src/main/scala/org/apache/spark/sql/SnappyBaseParser.scala [hunk 16]
```diff
@@ -294,7 +294,7 @@ object SnappyParserConsts {
   final val arithmeticOperator = CharPredicate('*', '/', '%', '&', '|', '^')
   final val exponent: CharPredicate = CharPredicate('e', 'E')
   final val numeric: CharPredicate = CharPredicate.Digit ++
-      CharPredicate('.')
+      CharPredicate('.') ++ exponent
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/SnappyBaseParser.scala]: revert

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 8 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Yes. Will change.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 5 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Not valid. Will remove.

File: cluster/src/test/scala/io/snappydata/QueryTest.scala (hunk 12 above)
Comment (rishitesh) [inline cluster/src/test/scala/io/snappydata/QueryTest.scala]: Yes. Merge issues.

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala (hunk 14 above)
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Input is at driver side. We have not tested driver side memory manager much. I will file a ticket and check if we need some more changes at driver side.

File: core/src/main/scala/org/apache/spark/sql/SnappyBaseParser.scala (hunk 16 above)
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/SnappyBaseParser.scala]: Yes.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 10 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Off-heap eviction at 90% seems fair, but IMO heap eviction at 90% causes a lot of activity during query execution, eviction and all creates a lot of garbage and it quickly reaches critical limit. I would suggest lets keep eviction at 80 and once we get more confidence we can change the limit.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 9 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Will change

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala (hunk 15 above)
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Hmm, yes, probably that would be better.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 6 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: Hmm , yes.. Will change.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 4 above)
Comment (rishitesh) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: not needed

File: core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala (hunk 15 above)
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Problem was, I wanted a memory allocation which should not evict data anymore. Right now our UMM acquireExecutionMemory() does not have that flag. I will have to change that before I can do this.

Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala]: Ok can let it remain storage for now but I think problem will be that under usual circumstances storage will always be about full and we don't want to evict data from here either.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 10 above)
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: My experience has been that GC always cleans up stuff before reaching 95%. So we can at least increase the critical to that especially if heap size is greater than say 4g (on Gem side too). Lets test more for eviction.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 17]
```diff
@@ -71,31 +74,47 @@ class SnappyUnifiedMemoryManager private[memory](
 
   private val onHeapStorageRegionSize = onHeapStorageMemoryPool.poolSize
 
-  private val maxHeapStorageSize = (maxHeapMemory *
-      conf.getDouble("spark.memory.storageMaxFraction", 0.9)).toLong
-  // TODO: [sumedh] Not being used?
-  val maxExecutionSize: Long = (maxHeapMemory * 0.75).toLong
+  val evictionFraction = SnappyUnifiedMemoryManager.getStorageEvictionPercent(conf)
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: This should really be a fraction so multiply with 0.01 right here

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 18]
```diff
@@ -580,6 +618,8 @@ object SnappyUnifiedMemoryManager extends Logging {
       math.max(getMaxHeapMemory / 20, 500L * 1024L * 1024L))
   }
 
+  private val DEFAULT_EVICTION_FRACTION = 80f
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: This should be a fraction so 0.8

File: core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala [hunk 19]
```diff
@@ -18,135 +18,85 @@ package org.apache.spark.memory
...
 
   override def acquireStorageMemoryForObject(objectName: String,
       blockId: BlockId,
       numBytes: Long,
       memoryMode: MemoryMode,
       buffer: UMMMemoryTracker,
       shouldEvict: Boolean): Boolean = {
-    logDebug(s"Acquiring mem [NO-OP] for $objectName $numBytes")
-    true
+    logDebug(s"Acquiring DefaultManager meemory for $objectName $numBytes")
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala]: typo: memory

File: core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala [hunk 20]
```diff
@@ -18,135 +18,85 @@ package org.apache.spark.memory
...
+    SparkEnv.get.memoryManager.acquireStorageMemory(blockId, numBytes, memoryMode)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/memory/StoreUnifiedManager.scala]: There might be some race conditions here that can cause NPE during startup/shutdown? Perhaps better to have SparkEnv.get.memoryManager as a val in the class.

**Comment by sumwale** [comment]
@rishitesh
Apart from some minor comments, the main remaining concern is whether eviction=false with acquireStorage memory will cause trouble. In many use-cases memory is almost full and this combination will likely fail frequently. I think that combination itself must be removed and instead all calling points should use eviction=false with acquireExecution memory.

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 21]
```diff
@@ -682,7 +675,7 @@ object SnappyUnifiedMemoryManager extends Logging {
     if (cache ne null) {
       val thresholds = cache.getResourceManager.getHeapMonitor.getThresholds
       if (thresholds.getEvictionThreshold > 0.1f) {
-        cache.getResourceManager.getHeapMonitor.getThresholds.getEvictionThreshold
+        (cache.getResourceManager.getHeapMonitor.getThresholds.getEvictionThreshold * 0.01).toFloat
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: change the method name to "getStorageEvictionFraction"

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala [hunk 22]
```diff
@@ -562,26 +562,20 @@ class SnappyUnifiedMemoryManager private[memory](
...
-    if (memoryForObject.addTo(key, -numBytes) >= 0L) {
-      super.releaseStorageMemory(numBytes, memoryMode)
-    } else {
-      // objectName was not present
-      memoryForObject.removeLong(key)
+    super.releaseStorageMemory(numBytes, memoryMode)
+    if (memoryForObject.containsKey(key)) {
+      if (memoryForObject.addTo(key, -numBytes) == numBytes) {
+        memoryForObject.removeLong(key)
+      }
```
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: I think original code was correct and more efficient. It avoids two lookup into the map for most cases. Something like below with your updated logic to always invoke super:

```scala
  super.releaseStorageMemory(numBytes, memoryMode)
  val originalNumBytes = memoryForObject.addTo(key, -numBytes)
  if (memoryForObject.addTo(key, -numBytes) <= numBytes) {
    // key not present or has gone down to zero
    memoryForObject.removeLong(key)
  }
```

File: cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala (hunk 17 above)
Comment (sumwale) [inline cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala]: rename getStorageEvictionPercent to getStorageEvictionFraction
